src/parse.rs

```diff
@@ -0,0 +1,641 @@
+use rug::{integer::IntegerExt64, Complete, Float, Integer};
+
+use crate::{
+    calculation_results::Number,
+    calculation_tasks::{CalculationBase, CalculationJob, INTEGER_CONSTRUCTION_LIMIT},
+    math::FLOAT_PRECISION,
+};
+
+const POI_STARTS: [char; 17] = [
+    '-', '!', '.', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '>', '&', '(', ')',
+];
+
+const PREFIX_OPS: [char; 1] = ['!'];
+const POSTFIX_OPS: [char; 2] = ['!', '?'];
+
+const INTEGER_ONLY_OPS: [i32; 1] = [-1];
+
+pub fn parse(mut text: &str, do_termial: bool) -> Vec<CalculationJob> {
+    // Parsing rules:
+    // - prefix has precedence before suffix (unimplemented)
+    // - anything within a spoiler should be ignored
+    // - operations may be nested through parentheses
+    // - operations can be negated through -
+    // - parens may contain:
+    //   - numbers
+    //   - operations
+    //   - parens
+    //   - whitespace
+    // - operations are:
+    //   - subfactorials !n
+    //   - (multi-)factorials n!+
+    //   - termials n?
+    // - numbers are in order:
+    //   - a string of digits
+    //   - a decimal separator and further digits
+    //   - a base 10 exponent, which is:
+    //     - an e or E followed by
+    //     - optionally a + or -
+    //     - a string of digits
+    // - numbers need to at least have the first or second criteria
+
+    // Parsing:
+    // 1. skip to interesting
+    // 2. If spoiler, skip
+    // 3. If negation, save
+    // 4. If paren start, push (with negation)
+    // 5. If paren end, pop
+    //   1. If had prefix, use base, set base
+    //   2. If has postfix, use base, set base
+    // 6. If prefix
+    //   1. If on number, set base
+    //     1. If has postfix, use base, set base
+    //   2. If on paren, push (with negation and level)
+    // 7. If number, parse
+    //   1. If has postfix, set base
+    //   2. If in parens, set as base
+    // 8. If on toplevel, add base to jobs
+    //
+    // when setting base:
+    // 1. If base is set, add previous to jobs
+    // 2. override base
+    let mut jobs = Vec::new();
+    let mut base: Option<CalculationBase> = None;
+    let mut paren_steps: Vec<(u32, Option<i32>)> = Vec::new();
+    let mut current_negative: u32 = 0;
+    while !text.is_empty() {
+        // Text (1.)
+        let Some(position_of_interest) = text.find(POI_STARTS) else {
+            break;
+        };
+        if position_of_interest != 0 {
+            current_negative = 0;
+        }
+        // so we can just ignore everything before
+        text = &text[position_of_interest..];
+        if text.starts_with(">!") {
+            // Spoiler (2.)
+            let end = text.find("!<").unwrap_or(1);
```

Comment:
To accept `>!5` as a factorial if there is no end tag
```suggestion
            let end = text.find("!<").unwrap_or(0);
```